src/tools/ScenarioMeasurement/Startup/PDNStartupParser.cs

```diff
@@ -0,0 +1,122 @@
+    // class PerfLabGenericEventSource : EventSource
+    // {
+    //     public static PerfLabGenericEventSource Log = new PerfLabGenericEventSource();
+    //     public void Startup() => WriteEvent(1);
+    // }
+
+    public class PDNStartupParser : IParser
+    {
+        public void EnableKernelProvider(ITraceSession kernel)
+        {
+            kernel.EnableKernelProvider(TraceSessionManager.KernelKeyword.Process, TraceSessionManager.KernelKeyword.Thread, TraceSessionManager.KernelKeyword.ContextSwitch);
+        }
+
+        public void EnableUserProviders(ITraceSession user)
+        {
+            user.EnableUserProvider("PaintDotnetTrace", TraceEventLevel.Verbose);
+        }
+
+        public IEnumerable<Counter> Parse(string mergeTraceFile, string processName, IList<int> pids, string commandLine)
+        {
+            var readyResults = new List<double>();
+            var startResults = new List<double>();
+            var idleResults = new List<double>();
+            var fullResults = new List<double>();
+            var threadTimes = new List<double>();
+            double threadTime = 0;
+            var ins = new Dictionary<int, double>();
+            double start = -1;
+            double rollingStart = -1;
+            int? pid = null;
+            using (var source = new ETWTraceEventSource(mergeTraceFile))
+            {
+
+                source.Kernel.ProcessStart += evt =>
+                {
+                    string commandLineArgs = commandLine.Substring(commandLine.LastIndexOf("\"")+1).Trim();
+                    string payloadCommandLineArgs = evt.CommandLine.Substring(evt.CommandLine.LastIndexOf("\"")+1).Trim();
+                    if (processName.Equals(evt.ProcessName, StringComparison.OrdinalIgnoreCase) && pids.Contains(evt.ProcessID) && payloadCommandLineArgs == commandLineArgs)
+                    {
+                        if (pid.HasValue)
+                        {
+                            // Processes might be reentrant. For now this traces the first (outermost) process of a given name.
+                            return;
+                        }
+                        pid = evt.ProcessID;
+                        start = evt.TimeStampRelativeMSec;
+                    }
+                };
+
+                source.Kernel.ThreadCSwitch += evt =>
+                {
+                    if (!pid.HasValue) // we're currently in a measurement interval
+                        return;
+
+                    if (evt.NewProcessID != pid && evt.OldProcessID != pid)
+                        return; // but this isn't our process
+
+                    if (evt.OldProcessID == pid) // this is a switch out from our process
+                    {
+                        if (ins.TryGetValue(evt.OldThreadID, out var value)) // had we ever recorded a switch in for this thread? 
+                        {
+                            threadTime += evt.TimeStampRelativeMSec - value;
```

Comment:
Hmm, that does not match my understanding or how the code has been operating for the last couple years. I talked with @brianrob who owns this code and he said that this behavior is single threaded. So I don't think that we need to worry about this here.

Reply:
ah I see, the docs said it was asynchronous, but maybe that is for when it is handling live events, rather than from parsing a file

Reply:
`ETWEventSource` is always single threaded and can only dispatch an event after the previous one has completed.  The events will be dispatched asynchronously from where they were produced (e.g. inside the producer process).  I suspect this is what the docs were referring to, but if you point me to where you saw this, I can make sure.

Reply:
This is the part of the docs that made me think it might be multi-threaded: [link](https://github.com/microsoft/perfview/blob/570cc0364c7e2407222f84600aebb9d9cb2f6007/documentation/TraceEvent/TraceEventProgrammersGuide.md#asynchronous)

Re-reading this part of the docs with your comment in mind made me realise my misunderstanding, so I think you are right in what part I was confused by. I mistakenly thought that since writing an event was a fire and forget operation that the event handlers themselves would also be called as part of the fire and forget, but missed the part about how the event is quickly written to a buffer.

Is there any possibility of events being received out-of-order during the time that the events are dispatched asynchronously to the buffer? Since the code above also assumes that all events are received in-order.

Reply:
Thanks for confirming - events can technically be written out-of-order when considering that there is one buffer per thread, and those buffers don't get flushed in time order.  You should not have a problem with this because you're capturing and then playing back the trace, and so the ETW trace reader API will handle keeping the events in time order.  The place where this can fall down is with real-time sessions, and in this case, there are collection options that can be used if this is absolutely critical.